### DISCUSSION TITLE
Value of the constraints at the solution

### DISCUSSION THREAD

**Discussion by joseph-gergaud** [JuliaSmoothOptimizers/Organization#36] [General]
We want to solve an optimization problem with contraints and we  would like for diagnostic to have the value of the constraints at the solution. Is it possible ?

**Comment by tmigot** [comment]
The returned structure by all the JuliaSmoothOptimizers is a `GenericExecutionStats`, see documentation [here `SolverCore.GenericExecutionStats`](https://juliasmoothoptimizers.github.io/SolverCore.jl/dev/reference/#SolverCore.GenericExecutionStats).

The particular field you might be interested in are:
```
- `dual_feas`: The dual feasibility norm at solution (default: Inf);
- `primal_feas`: The primal feasibility norm at solution (default: 0.0 if unconstrained, Inf otherwise);
```

**Reply by joseph-gergaud** [reply]
Yes, I have seen this, but we don't have de value of the constraints (Ipopt return these values)

**Reply by abelsiqueira** [reply]
You should be able to evaluate the constraints value at the solution with something like

```julia
using NLPModels
cx = cons(nlp, output.solution)
```

The `NLPModels` defines the API, `cons` is the function that evaluates constraints, `nlp` is the problem and `output` is the object returned by the solver.

**Comment by joseph-gergaud** [comment]
When I execute the following program

```julia
using NLPModelsIpopt, ADNLPModels
nlp = ADNLPModel(x -> sum(x.^2), ones(3));

f(x) = 0.5*((x[1]-1)^2 + x[2]^2)
c(x) = [-x[1]+x[2]^2]
x0 = [1.;1.]
println("f(x0) = ", f(x0))
nlp = ADNLPModel(f, x0, c, [0.], [0.])

output = ipopt(nlp, print_level = 5)


println("solution = ", output.solution)

println("cons(nlp, output.solution) = ", cons(nlp, output.solution))
```
I obtain : 

EXIT: Optimal Solution Found.
solution = [0.5000000001101497, 0.7071067813550226]
ERROR: LoadError: UndefVarError: cons not defined
Stacktrace:
 [1] top-level scope
   @ ~/control-toolbox/CTDirect.jl/work/toto.jl:15
 [2] include(fname::String)
   @ Base.MainInclude ./client.jl:476
 [3] top-level scope
   @ REPL[6]:1
in expression starting at /Users/gergaud/control-toolbox/CTDirect.jl/work/toto.jl:15

And if write : 
```julia
println("ADNLPModels.cons(nlp, output.solution) = ", ADNLPModels.cons(nlp, output.solution))
```
It's ok I obtain

EXIT: Optimal Solution Found.
solution = [0.5000000001101497, 0.7071067813550226]
ADNLPModels.cons(nlp, output.solution) = [1.2811007810142883e-10]

I think you have to export the fonction cons

Thanks,

**Reply by abelsiqueira** [reply]
Yes, the `cons` function is exported only in `NLPModels`, so you would have to `using NLPModels` to access it or qualify it as you have done.
Either way, I'm glad it works. Let us know if you need more help :+1:

**Comment by joseph-gergaud** [comment]
Thanks a lot. We can close the discussion
